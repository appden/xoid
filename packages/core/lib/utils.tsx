import { Create, Listener, Store } from "./types"

export type Meta = {
  parentMeta?: Meta
  node: any
  cache: Record<string, any>
  key?: string
  shape?: any
  root: ReturnType<typeof createRoot>
}

//@ts-ignore
export const RECORD = window.__XOID_SYMBOLS__ ? window.__XOID_SYMBOLS__.record : Symbol()
//@ts-ignore
export const META = window.__XOID_SYMBOLS__ ? window.__XOID_SYMBOLS__.meta : Symbol()
//@ts-ignore
if(!window.__XOID_SYMBOLS__) window.__XOID_SYMBOLS__ = { record: RECORD, meta: META }

export const createCell = (parentMeta: Meta, key: string): any => {
  const node = parentMeta.node[key]
  if (node && node[META]) return node
  if (parentMeta.cache[key]) return parentMeta.cache[key]

  const meta = {
    parentMeta,
    key,
    get node() {
      return parentMeta.node[key]
    },
    set node(value) {
      if (meta.root.mutable) {
        parentMeta.node[key] = value
        return
      }
      const copy = generateNew(parentMeta.node)
      copy[key] = value
      parentMeta.node = copy
    },
    cache: {},
    root: parentMeta.root,
    shape: parentMeta.shape && (parentMeta.shape[RECORD] || parentMeta.shape[key]),
  } as Meta

  const proxy: any = new Proxy(createTarget(meta), {
    get(_, prop: string | symbol) {
      if (prop === META) return meta

      // start: prototype stuff
      const node = meta.node
      if ((prop as symbol) === Symbol.toPrimitive) return () => node
      if (!Object.prototype.hasOwnProperty.call(node, prop)) {
        if (node[prop]) {
          return node[prop].bind
            ? node[prop].bind(createCell(meta.parentMeta as Meta, meta.key as string))
            : node[prop]
        }
      }
      // end: prototype stuff

      return createCell(meta, prop as string)
    },
    set() {
      return false
    },
    has(_, key) {
      return key in meta.node
    },
    ownKeys() {
      return Reflect.ownKeys(meta.node)
    },
    getOwnPropertyDescriptor() {
      return {
        enumerable: true,
        configurable: true,
      }
    },
  })
  parentMeta.cache[key] = proxy
  return proxy
}

const generateNew = (obj: any) => Object.assign(Array.isArray(obj) ? [] : Object.create(obj), obj)

export const createTarget = (meta: Meta) => {
  return function (input?: any) {
    if (arguments.length === 0) return meta.node
    const newValue = typeof input === 'function' ? input(meta.node) : input
    if (meta.node === newValue) return
    meta.node = newValue
    meta.root.notify()
  }
}

export const createRoot = (mutable?: boolean) => {
  const listeners = new Set<() => void>()
  const notify = () => listeners.forEach((listener) => listener())
  const subscribe = (listener: () => void) => {
    listeners.add(listener)
    return () => listeners.delete(listener)
  }
  return { listeners, notify, subscribe, mutable }
}

export const createInstance = (shape?: any): Create => (init?: any, mutable?: any) => {
  const isFunction = typeof init === 'function'
  const root = createRoot(mutable)
  const store = createCell(
    {
      node: { value: isFunction ? null : init },
      cache: {},
      root,
      shape: { value: shape },
    },
    'value'
  )
  if (isFunction) {
    const unsubs = new Set<() => void>()
    const getter = (store: Store<any>) => {
      unsubs.add(subscribe(store, updateState))
      return store()
    }
    const updateState = () => {
      unsubs.forEach((fn) => fn())
      unsubs.clear()
      return store((init as Function)(getter))
    }
    updateState()
  }
  return store
}

/**
 * Subscribes to a store, or a partial store.
 * @see [xoid.dev/docs/api/subscribe](https://xoid.dev/docs/api/subscribe)
 */

 export const subscribe = <T extends any>(store: Store<T>, fn: Listener<T>): (() => void) => {
  let prevValue = store()
  return ((store as any)[META] as Meta).root.subscribe(() => {
    const nextValue = store()
    if (nextValue !== prevValue) fn(nextValue as any)
    prevValue = nextValue
  })
}
