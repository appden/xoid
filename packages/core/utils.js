'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

//@ts-ignore
var RECORD = window.__XOID_SYMBOLS__ ? window.__XOID_SYMBOLS__.record : Symbol();
//@ts-ignore
var META = window.__XOID_SYMBOLS__ ? window.__XOID_SYMBOLS__.meta : Symbol();
//@ts-ignore
if (!window.__XOID_SYMBOLS__)
    window.__XOID_SYMBOLS__ = { record: RECORD, meta: META };
var createCell = function (parentMeta, key) {
    var node = parentMeta.node[key];
    if (node && node[META])
        return node;
    if (parentMeta.cache[key])
        return parentMeta.cache[key];
    var meta = {
        parentMeta: parentMeta,
        key: key,
        get node() {
            return parentMeta.node[key];
        },
        set node(value) {
            if (meta.root.mutable) {
                parentMeta.node[key] = value;
                return;
            }
            var copy = generateNew(parentMeta.node);
            copy[key] = value;
            parentMeta.node = copy;
        },
        cache: {},
        root: parentMeta.root,
        shape: parentMeta.shape && (parentMeta.shape[RECORD] || parentMeta.shape[key]),
    };
    var proxy = new Proxy(createTarget(meta), {
        get: function (_, prop) {
            if (prop === META)
                return meta;
            // start: prototype stuff
            var node = meta.node;
            if (prop === Symbol.toPrimitive)
                return function () { return node; };
            if (!Object.prototype.hasOwnProperty.call(node, prop)) {
                if (node[prop]) {
                    return node[prop].bind
                        ? node[prop].bind(createCell(meta.parentMeta, meta.key))
                        : node[prop];
                }
            }
            // end: prototype stuff
            return createCell(meta, prop);
        },
        set: function () {
            return false;
        },
        has: function (_, key) {
            return key in meta.node;
        },
        ownKeys: function () {
            return Reflect.ownKeys(meta.node);
        },
        getOwnPropertyDescriptor: function () {
            return {
                enumerable: true,
                configurable: true,
            };
        },
    });
    parentMeta.cache[key] = proxy;
    return proxy;
};
var generateNew = function (obj) { return Object.assign(Array.isArray(obj) ? [] : Object.create(obj), obj); };
var createTarget = function (meta) {
    return function (input) {
        if (arguments.length === 0)
            return meta.node;
        var newValue = typeof input === 'function' ? input(meta.node) : input;
        if (meta.node === newValue)
            return;
        meta.node = newValue;
        meta.root.notify();
    };
};
var createRoot = function (mutable) {
    var listeners = new Set();
    var notify = function () { return listeners.forEach(function (listener) { return listener(); }); };
    var subscribe = function (listener) {
        listeners.add(listener);
        return function () { return listeners.delete(listener); };
    };
    return { listeners: listeners, notify: notify, subscribe: subscribe, mutable: mutable };
};
var createInstance = function (shape) { return function (init, mutable) {
    var isFunction = typeof init === 'function';
    var root = createRoot(mutable);
    var store = createCell({
        node: { value: isFunction ? null : init },
        cache: {},
        root: root,
        shape: { value: shape },
    }, 'value');
    if (isFunction) {
        var unsubs_1 = new Set();
        var getter_1 = function (store) {
            unsubs_1.add(subscribe(store, updateState_1));
            return store();
        };
        var updateState_1 = function () {
            unsubs_1.forEach(function (fn) { return fn(); });
            unsubs_1.clear();
            return store(init(getter_1));
        };
        updateState_1();
    }
    return store;
}; };
/**
 * Subscribes to a store, or a partial store.
 * @see [xoid.dev/docs/api/subscribe](https://xoid.dev/docs/api/subscribe)
 */
var subscribe = function (store, fn) {
    var prevValue = store();
    return store[META].root.subscribe(function () {
        var nextValue = store();
        if (nextValue !== prevValue)
            fn(nextValue);
        prevValue = nextValue;
    });
};

exports.META = META;
exports.RECORD = RECORD;
exports.createCell = createCell;
exports.createInstance = createInstance;
exports.createRoot = createRoot;
exports.createTarget = createTarget;
exports.subscribe = subscribe;
