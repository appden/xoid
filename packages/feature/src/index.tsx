type Config<T, U> = {
  id?: string | symbol
  defaultOptions?: T
  options?: U
}

export type From = {
  <U extends new (...args: any) => any>(item: U): InstanceType<U>
  <U extends { id: any; type: any }>(item: U): U['type']
}

export abstract class FeatureAbstract<Options, DefaultOptions> {
  externalOptions!: Options
  options!: DefaultOptions & Options
  from!: From
  main?: () => void
}

export type FeatureConstructor = new (options: any) => FeatureAbstract<any, any>

type PrefilledFeature<T> = T extends new (options: any) => infer Q
  ? new (options: {}) => Omit<Q, 'externalOptions'> & { externalOptions: {} }
  : never

export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never

type GetKey<I, K extends string | symbol | number> = I extends {
  [L in K]: infer P
}
  ? P
  : never

export type ExtractKey<U, K extends string | number | symbol> = GetKey<UnionToIntersection<U>, K>

type MergedTypes<P extends FeatureConstructor> = {
  [K in keyof UnionToIntersection<InstanceType<P>>]: ExtractKey<InstanceType<P>, K>
}

const REFERENCE = Symbol()

const defaultFrom: any = () => console.warn('Please use it inside compose')

const current = { from: defaultFrom }

const createFrom = (contextMap: any) =>
  function (this: any, input: any) {
    const output = contextMap.get(typeof input === 'object' ? input.id : input)
    if (!output) {
      console.error(`Dependency '${input.name}' not found in '${this.constructor.name}'.`)
    }
    return output
  }

const interceptFrom = <T,>(nextFrom: any, fn: () => T) => {
  current.from = nextFrom
  const ans = fn()
  current.from = defaultFrom
  return ans
}

export function Feature<T extends Config<any, any> = { defaultOptions: {}; options: {} }>(
  config = {
    id: Symbol(),
    defaultOptions: {},
    options: {},
  } as T
) {
  return class {
    static id = config.id

    externalOptions!: T['options']
    options!: T['options'] & T['defaultOptions']
    from!: From

    constructor(options: T['options']) {
      this.from = (a: any) => current.from(a)
      this.options = { ...config.defaultOptions, ...options }
    }
  }
}

export const composeFeatures =
  <P extends FeatureConstructor, V>(
    config: P[],
    afterAll: (from: From, common: MergedTypes<P>) => V
  ) =>
  (options: ExtractKey<InstanceType<P>, 'externalOptions'>) => {
    const contextMap = new Map<P, InstanceType<P>>()
    config
      .map((item) => {
        const instance = runInner(contextMap)(item, options)
        // @ts-ignore
        contextMap.set(item[REFERENCE] || item, instance)
        // @ts-ignore
        if (item.id) {
          // @ts-ignore
          contextMap.set(item.id, instance)
        }
        return instance
      })
      .forEach((instance) => {
        instance?.main?.()
      })
    const ans = afterAll(contextMap.get.bind(contextMap) as From, {} as MergedTypes<P>)
    return ans
  }

const runInner =
  (contextMap: Map<FeatureConstructor, FeatureAbstract<any, any>>) =>
  <P extends FeatureConstructor>(
    item: P,
    options: ExtractKey<InstanceType<P>, 'externalOptions'>
  ) => {
    const from = createFrom(contextMap)
    const instance = interceptFrom(from, () => new item(options))
    instance.from = from
    return instance
  }

export const prefill = <
  T extends FeatureConstructor,
  U extends ExtractKey<InstanceType<T>, 'externalOptions'>
>(
  item: T,
  options: U | ((from: From) => U)
) =>
  // @ts-ignore
  class extends item {
    static [REFERENCE] = item
    constructor() {
      // @ts-ignore
      super(typeof options === 'function' ? options(current.from) : options)
    }
  } as PrefilledFeature<T>
