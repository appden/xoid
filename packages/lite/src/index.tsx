export { create } from '../../xoid/src/internal/lite'
