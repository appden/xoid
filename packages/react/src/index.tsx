import { useSyncExternalStore } from 'use-sync-external-store/shim'
import React, { useEffect, useLayoutEffect, useRef, useDebugValue } from 'react'
import { Atom, Usable, create, use as _use } from 'xoid'

export type Pair<T> = { value: T; onChange: (value: T) => void }

export type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>

export type PropsOf<T> = T extends ComponentType
  ? React.ComponentProps<T>
  : T extends (...args: any) => any
  ? Parameters<T>[0]
  : never

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
export const useIsoLayoutEffect = window === undefined ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

const useReactAdapter = () => {
  const result = useConstant(() => {
    const mount = (_use as any).createEvent()
    const unmount = (_use as any).createEvent()
    return {
      mount,
      unmount,
      adapter: {
        mount: mount.add,
        unmount: unmount.add,
      },
    }
  })
  useIsoLayoutEffect(() => {
    result.mount.fire()
    return result.unmount.fire
  }, [])
  return result.adapter
}

export type ReactAdapter = ReturnType<typeof useReactAdapter>

/**
 * Subscribes to an atom inside a React function component.
 *
 * An atom, or a function returning an atom can be passed as the first argument.
 * When the second optional argument is set to `true`, it will also consume the usables of the atom.
 * @see [xoid.dev/docs/api-react/use-atom](https://xoid.dev/docs/api-react/use-atom)
 */
export function useAtom<T>(atom: Atom<T>): T
export function useAtom<T>(atom: () => Atom<T>): T
export function useAtom<T, U>(atom: Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(atom: () => Atom<T> & Usable<U>, use: true): [T, U]
export function useAtom<T, U>(maybeAtom: Atom<T> | (() => Atom<T>), use?: boolean): [T, U] | T {
  const atom = useConstant(() => (typeof maybeAtom === 'function' ? maybeAtom() : maybeAtom))
  const value = useSyncExternalStore(
    atom.subscribe,
    () => atom.value,
    () => atom.value
  )
  useDebugValue(value)
  return use ? ([value, _use(atom as any)] as [T, U]) : (value as T)
}

/**
 * Can be used to create local state inside React components. Similar to `React.useMemo`,
 * but creates values **exactly once**.
 * @see [xoid.dev/docs/api-react/use-setup](https://xoid.dev/docs/api-react/use-setup)
 */
export function useSetup<T>(fn: ($props: undefined, adapter: ReactAdapter) => T): T
export function useSetup<T, P>(fn: ($props: Atom<P>, adapter: ReactAdapter) => T, props: P): T
export function useSetup(fn: ($props: any, adapter: any) => any, props?: any): any {
  /* eslint-disable react-hooks/rules-of-hooks */
  // Calling hooks conditionally wouldn't be an issue here, because we rely on just
  // the Function.length and Arguments.length, which will remain static.
  const api = fn.length > 1 ? useReactAdapter() : undefined
  if (arguments.length > 1) {
    const [$props, result] = useConstant(() => {
      const $props = create(() => props)
      const result = fn($props, api)
      return [$props, result] as const
    })
    useIsoLayoutEffect(() => ($props as Atom<any>).set(props), [props])
    return result
  } else {
    const result = useConstant(() => fn(undefined, api))
    return result
  }
  /* eslint-enable react-hooks/rules-of-hooks */
}
