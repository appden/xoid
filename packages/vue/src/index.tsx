import { getCurrentScope, onScopeDispose, shallowRef } from 'vue'
import { Atom } from 'xoid'

export function useAtom<T>(atom: Atom<T>) {
  const state = shallowRef()

  const unsubscribe = atom.subscribe((value) => {
    state.value = value
  })

  getCurrentScope() && onScopeDispose(unsubscribe)
  return state
}
