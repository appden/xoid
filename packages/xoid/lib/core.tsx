import {
  createNotifier,
  createSelector,
  createTarget,
  META,
  USEABLE,
  Init,
  Atom,
} from '@xoid/engine'
import { select } from './lens'

const useable = Symbol()
export type Useable<U> = { [useable]: U }

/**
 * Gets the "useable" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/api/use)
 */
export function use<T extends unknown, U>(atom: Atom<T>, fn: (state: T) => U): Atom<U>
export function use<T extends unknown, U extends keyof T>(atom: Atom<T>, fn: U): Atom<T[U]>
export function use<T extends any>(atom: Useable<T>): T
export function use(atom: any, fn?: any): any {
  if (!fn) return (atom as any)[USEABLE]
  return select(atom, fn)
}

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used too attach "useable"s to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/api/create)
 */

export function create<T>(): Atom<T | undefined>
export function create<T>(init: Init<T>): Atom<T>
export function create<T, U>(init: Init<T>, useable?: (atom: Atom<T>) => U): Atom<T> & Useable<U>
export function create<T, U = undefined>(init?: Init<T>, useable?: (atom: Atom<T>) => U): Atom<T> {
  const meta = { notifier: createNotifier(), node: init }
  const target = createTarget(meta)
  if (typeof init === 'function') createSelector(target as unknown as Atom<T>, init as Function)
  ;(target as any)[META] = meta
  if (useable && typeof useable === 'function') (target as any)[USEABLE] = useable(target as any)
  return target as any
}
