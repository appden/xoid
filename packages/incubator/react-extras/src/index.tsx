import { useAtom, useSetup } from '@xoid/react'
export { slice } from './slice'

/**
 * Can be used to consume a **value-onChange pair** as an atom inside React components.
 *
 * Third argument, when set to `true`, enables optimistic updates. This can be
 * used for keeping a synced local state, and debouncing the updates sent to the parent.
 * @see [xoid.dev/docs/api-react/use-thru](https://xoid.dev/docs/api-react/use-thru)
 */
export const useThru = <T,>(value: T, onChange?: (value: T) => void, optimistic = false) => {
  const atom = useSetup(
    ($props) => {
      const atom = $props.focus('value')
      const props = $props.value
      const prevSet = atom.set
      atom.set = (value: T) => {
        if (props.optimistic) prevSet(value)
        props.onChange?.(value)
      }
      return atom
    },
    { value, onChange, optimistic }
  )
  useAtom(atom)
  return atom
}
