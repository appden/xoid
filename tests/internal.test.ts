import { createStore, subscribe, set, get } from '../src'
import { deepClone, destroy, memberMap } from '../src/core/utils'

it('can destroy the store', () => {
  const store = createStore({
    value: 1,
  })

  subscribe(store, () => {
    throw new Error('did not clear listener on destroy')
  })
  destroy(store)

  set(store, { value: 2 })
  expect(get(store.value)).toEqual(2)
})

it('clones the state to produce two trees', () => {
  const value = {
    alpha: 3,
    beta: 5,
    alphaFn: (a: 5, b: 7) => a + b,
    deeply: {
      nested: { value: 'value1' },
      otherNested: {
        value1: 'value1',
        value2: 'value2',
        value3: 'value3',
      },
    },
  }

  const clones = deepClone(value, { mockStore: 'mockStore' } as any)
  expect(clones).toMatchSnapshot()

  const [symbolic, normalized] = clones

  function traverse(o: any, func: any) {
    for (var i in o) {
      func(i, o[i])
      if (o[i] !== null && typeof o[i] == 'object') {
        //going one step down in the object tree!!
        traverse(o[i], func)
      }
    }
  }
  const adresses: any = []
  traverse(symbolic, (key: any, value: any) => {
    if (key !== '=')
      adresses.push({ key, address: memberMap.get(value)?.address })
  })
  expect(adresses).toMatchSnapshot()
})
