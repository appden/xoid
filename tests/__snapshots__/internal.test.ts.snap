// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clones the state to produce two trees 1`] = `
Array [
  Object {
    "alpha": Object {
      "=‎‎": 3,
    },
    "alphaFn": [Function],
    "beta": Object {
      "=‎‎": 5,
    },
    "deeply": Object {
      "nested": Object {
        "value": Object {
          "=‎‎": "value1",
        },
      },
      "otherNested": Object {
        "value1": Object {
          "=‎‎": "value1",
        },
        "value2": Object {
          "=‎‎": "value2",
        },
        "value3": Object {
          "=‎‎": "value3",
        },
      },
    },
  },
  Object {
    "alpha": 3,
    "alphaFn": [Function],
    "beta": 5,
    "deeply": Object {
      "nested": Object {
        "value": "value1",
      },
      "otherNested": Object {
        "value1": "value1",
        "value2": "value2",
        "value3": "value3",
      },
    },
  },
  Set {},
]
`;

exports[`clones the state to produce two trees 2`] = `
Array [
  Object {
    "address": Array [
      "alpha",
    ],
    "key": "alpha",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
  Object {
    "address": Array [
      "beta",
    ],
    "key": "beta",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
  Object {
    "address": Array [
      "alphaFn",
    ],
    "key": "alphaFn",
  },
  Object {
    "address": Array [
      "deeply",
    ],
    "key": "deeply",
  },
  Object {
    "address": Array [
      "nested",
    ],
    "key": "nested",
  },
  Object {
    "address": Array [
      "value",
    ],
    "key": "value",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
  Object {
    "address": Array [
      "otherNested",
    ],
    "key": "otherNested",
  },
  Object {
    "address": Array [
      "value1",
    ],
    "key": "value1",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
  Object {
    "address": Array [
      "value2",
    ],
    "key": "value2",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
  Object {
    "address": Array [
      "value3",
    ],
    "key": "value3",
  },
  Object {
    "address": undefined,
    "key": "=‎‎",
  },
]
`;
